snakemake/executors.py

```diff
@@ -2052,3 +2059,605 @@ def run_wrapper(
+        if self._save_storage_cache:
+            logger.debug("Requested to save workflow cache, skipping cleanup.")
+        else:
+            for package in self._build_packages:
+                blob = self.bucket.blob(package)
+                if blob.exists():
+                    logger.debug("Deleting blob %s" % package)
+                    blob.delete()
+
+        # perform additional steps on shutdown if necessary
+        super().shutdown()
+
+    def cancel(self):
+        """cancel execution, usually by way of control+c. Cleanup is done in
+           shutdown (deleting cached workdirs in Google Cloud Storage
+        """
+        import googleapiclient
+
+        # projects.locations.operations/cancel
+        operations = self._api.projects().locations().operations()
+
+        for job in self.active_jobs:
+            request = operations.cancel(name=j.jobname)
+            logger.debug("Cancelling operation {}".format(j.jobid))
+            try:
+                status = self._retry_request(request)
+            except (Exception, BaseException, googleapiclient.errors.HttpError) as ex:
+                continue
+
+        self.shutdown()
+
+    def _generate_job_resources(self, job):
+        """given a particular job, generate the resources that it needs,
+           including default regions and the virtual machine configuration
+        """
+        # Right now, do a best effort mapping of resources to instance types
+        cores = job.resources.get("_cores", 1)
+        nodes = job.resources.get("_nodes", 1)
+        mem_mb = job.resources.get("mem_mb", 100)
+
+        # Regular expression to determine if zone in region
+        regexp = "^(%s)" % "|".join(self.regions)
+
+        # Retrieve zones, filter down to selected regions
+        zones = self._retry_request(
+            self._compute_cli.zones().list(project=self.project)
+        )
+        zones = [z for z in zones["items"] if re.search(regexp, z["name"])]
+
+        # Retrieve machine types available across zones
+        # https://cloud.google.com/compute/docs/regions-zones/
+        lookup = {}
+        for zone in zones:
+            request = self._compute_cli.machineTypes().list(
+                project=self.project, zone=zone["name"]
+            )
+            lookup[zone["name"]] = self._retry_request(request)["items"]
+
+        # Only keep those that are shared, use last zone as a base
+        machineTypes = {mt["name"]: mt for mt in lookup[zone["name"]]}
+        del lookup[zone["name"]]
+
+        # Update final list based on the remaining
+        to_remove = set()
+        for zone, types in lookup.items():
+            names = [x["name"] for x in types]
+            for machineType in list(machineTypes.keys()):
+                if machineType not in names:
+                    to_remove.add(machineType)
+
+        for machineType in to_remove:
+            del machineTypes[machineType]
```

Comment:
Without digging too much into the details here (I might be wrong), my feeling is that this could be a one-liner instead.

Reply:
Maybe not a one-liner but at least less dict editing? Rule of thumb, try to replace `del` statements with better initial filtering. Try to replace nested loops with comprehensions.

Reply:
One option to get this entire method cleaner would be to decompose it into smaller "filters" that are subsequently applied to the list or dict of machines?